akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala

```diff
@@ -405,60 +409,45 @@ private[akka] class PersistentShard(
     }
   }
 
-  private def retrieveEntityRecoveryConfig(): Config = {
-    val path = settings.tuningParameters.entityRecoveryStrategyConfigPath
-    if (path.isEmpty)
-      ConfigFactory.empty()
-    else
-      context.system.settings.config.getConfig(path)
+  private def restartRememberedEntities(): Unit = {
+    rememberedEntitiesRecoveryStrategy.recoverEntities(state.entities).foreach { scheduledRecovery ⇒
+      import context.dispatcher
+      scheduledRecovery.filter(_.nonEmpty).map(RestartEntities).pipeTo(self)
+    }
   }
 }
 
-abstract class EntityRecoveryStrategy {
-  import ShardRegion.EntityId
-  import akka.actor.ActorContext
-  def recoverEntities(context: ActorContext, entities: Set[EntityId])
+object EntityRecoveryStrategy {
+  def allStrategy(): EntityRecoveryStrategy = new AllAtOnceEntityRecoveryStrategy()
+
+  def constantStrategy(actorSystem: ActorSystem, frequency: FiniteDuration, numberOfEntities: Int): EntityRecoveryStrategy =
+    new ConstantRateEntityRecoveryStrategy(actorSystem, frequency, numberOfEntities)
 }
 
-final class AllAtOnceEntityRecoveryStrategy extends EntityRecoveryStrategy {
+trait EntityRecoveryStrategy {
   import ShardRegion.EntityId
-  import akka.actor.ActorContext
-  import Shard.RestartEntities
-  override def recoverEntities(context: ActorContext, entities: Set[EntityId]): Unit =
-    if (entities.nonEmpty)
-      context.self ! RestartEntities(entities)
+  import scala.concurrent.Future
+
+  def recoverEntities(entities: Set[EntityId]): Set[Future[Set[EntityId]]]
```

Comment:
@ktoso had to change the interface a bit from what Patrik was suggesting so that a future could be returned for each scheduled increment of the recovery.